### ISSUE TITLE
NNFramework

### ISSUE THREAD

**Issue by dmccloskey** [dmccloskey/EvoNet#18]
### Description
Barebones neural network framework

### Objectives
- [ ] SGS with adaptive and random learning rate
- [ ] Cost functions (e.g., 1/2(delta)**2)
- [ ] Mini batch with average update
- [ ] Foward propogation using an adjacency graph
- [ ] Backward propogation using an adjacency graph
- [ ] Node definitions (e.g., ReLU)
- [ ] Integration with Eigen

### Objectives (Bonus)
- [ ] GPU distribution

### Test
- Autoencoder

### Notes
- biases are weights with unitary output (i.e., w*o where o = 1.0)
- apply a function to an eigen matrix: https://stackoverflow.com/questions/33786662/apply-function-to-all-eigen-matrix-element
- eigen tensor: http://eigen.tuxfamily.org/index.php?title=Tensor_support,
